### PR TITLE
Perform tiny improvements of backends

### PR DESCRIPTION
**Pull request by mrkn**
I'd like to make backends improve little.

1. Stop using "plotter adapter", keep using "backend".
2. Move backend classes under `Charty::Backends` module namespace.
3. Remove `Name` constants from backend classes.
4. Alter the algorithm for searching a backend by using a hash table.

Use the following link to see the diff without whitespace changes:
https://github.com/red-data-tools/charty/pull/35/files?w=1

### CODE REVIEW

File: lib/charty/backends/pyplot.rb [hunk 1]
```diff
@@ -2,109 +2,109 @@
...
 
 module Charty
-  class PyPlot < PlotterAdapter
-    Name = "pyplot"
-
-    def initialize
-      @plot = Matplotlib::Pyplot
-    end
+  module Backends
+    class Pyplot < Base
```
Comment (284km) [inline lib/charty/backends/pyplot.rb]: I see, I know that it's officially Pyplot, not PyPlot 👍

Comment (mrkn) [reply lib/charty/backends/pyplot.rb]: I needed to rename it to let `Backends.register_backend` generate `"pyplot"` from the class name.

Comment (284km) [reply lib/charty/backends/pyplot.rb]: I got it.

File: lib/charty/backends.rb [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+      end
+      require "charty/backends/#{backend_name}"
+      backend_class = @backends[backend_name]
+      raise BackendNotLoadedError.new("Backend for '#{backend_name}' is not found.") unless backend_class
+      backend_class.new
+    end
+
+    class Base
+      def self.inherited(backend_class)
+        Charty::Backends.register_backend(backend_class)
```
Comment (kou) [inline lib/charty/backends.rb]: We use "calling `register` explicitly API" in table adapters. How about using the same API in backends?

File: lib/charty/backends.rb [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  module Backends
+    def self.register_backend(backend_class)
+      @backends ||= {}
+      key = backend_class.name[/(?:::)?(\w+)\z/, 1]
+      key.gsub!(/\A([A-Z])/) { $1.downcase }
+      key.gsub!(/([a-z\d])([A-Z])/, '\1_\2')
+      @backends[key] = backend_class
+    end
+
+    def self.create(backend_name)
```
Comment (kou) [inline lib/charty/backends.rb]: We use "finding class API" instead of "creating object API" in table adapters. How about using the same APi in backends?

File: lib/charty/backends.rb [hunk 4]
```diff
@@ -0,0 +1,32 @@
+module Charty
+  class BackendNotLoadedError < RuntimeError; end
+
+  module Backends
+    def self.register_backend(backend_class)
```
Comment (kou) [inline lib/charty/backends.rb]: We can omit `_backend` because this is `Backends` module.

File: lib/charty/backends.rb (hunk 2 above)
Comment (mrkn) [reply lib/charty/backends.rb]: Indeed.  I'll do.

File: lib/charty/backends.rb (hunk 3 above)
Comment (mrkn) [reply lib/charty/backends.rb]: Indeed.  I'll do, too.

File: lib/charty/backends.rb (hunk 2 above)
Comment (mrkn) [reply lib/charty/backends.rb]: Done

File: lib/charty/backends.rb (hunk 4 above)
Comment (mrkn) [reply lib/charty/backends.rb]: Done

File: lib/charty/backends.rb (hunk 3 above)
Comment (mrkn) [reply lib/charty/backends.rb]: Done

**Comment by mrkn** [comment]
I've made fixes for @kou's comments, and added some commits for the following modifications.

- Taking care of `LoadError` in `find_backend_class`
- Making loading backend libraries lazy, and preloading bundled backends

**Comment by kou** [comment]
Thanks!
Ignoring `LoadError` without notifying users will be difficult to debug.
I'll push some commits to improve this case.

**Comment by kou** [comment]
How about 35d26c4 ?
It always register all backends. So `Charty::Backends.names` works without any `Charty::Plotter.new(...)`.
It adds `prepare` hook to each backend class. Normally, each backend class loads dependency libraries. If loading dependency libraries is failed, `Charty::BackendLoadError` is raised. Users can know details by checking the `caused` error.

File: lib/charty/backends.rb [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+    def self.register(name, backend_class)
+      @backends[name] = {
+        class: backend_class,
+        prepared: false,
+      }
+    end
+
+    def self.find_backend_class(name)
+      backend = @backends[name]
```
Comment (kou) [inline lib/charty/backends.rb]: Should we normalize `name` to `String`? Can we always use `Symbol`?

File: test/plotter_test.rb [hunk 6]
```diff
@@ -3,8 +3,8 @@
 class PlotterTest < Test::Unit::TestCase
   def setup
     @plotter = Charty::Plotter.new(:pyplot)
-    assert_instance_of(Charty::PyPlot,
-                       @plotter.instance_variable_get(:@plotter_adapter))
+    assert_instance_of(Charty::Backends::Pyplot,
+                       @plotter.instance_variable_get(:@backend))
```
Comment (kou) [inline test/plotter_test.rb]: I think that we can remove this assertion. But this is out of scope for this pull request.

**Comment by mrkn** [comment]
> How about 35d26c4 ?

This commit is almost the same as what I'll create a new pull-request after this will be merged:

```diff
diff --git a/lib/charty/backends.rb b/lib/charty/backends.rb
index fc58470..63a3869 100644
--- a/lib/charty/backends.rb
+++ b/lib/charty/backends.rb
@@ -36,3 +36,10 @@ module Charty
     end
   end
 end
+
+require_relative 'backends/bokeh'
+require_relative 'backends/google_chart'
+require_relative 'backends/gruff'
+require_relative 'backends/plotly'
+require_relative 'backends/pyplot'
+require_relative 'backends/rubyplot'
diff --git a/lib/charty/backends/bokeh.rb b/lib/charty/backends/bokeh.rb
index def626d..0b9dec8 100644
--- a/lib/charty/backends/bokeh.rb
+++ b/lib/charty/backends/bokeh.rb
@@ -1,9 +1,8 @@
-require 'pycall'
-
 module Charty
   module Backends
     class Bokeh
       def initialize
+        require 'pycall'
         @plot = PyCall.import_module('bokeh.plotting')
       end
 
diff --git a/lib/charty/backends/gruff.rb b/lib/charty/backends/gruff.rb
index 5d7e8a4..f952f3a 100644
--- a/lib/charty/backends/gruff.rb
+++ b/lib/charty/backends/gruff.rb
@@ -1,10 +1,10 @@
-require 'gruff'
 require 'fileutils'
 
 module Charty
   module Backends
     class Gruff
       def initialize
+        require 'gruff'
         @plot = ::Gruff
       end
 
diff --git a/lib/charty/backends/pyplot.rb b/lib/charty/backends/pyplot.rb
index d44eed6..1b923a2 100644
--- a/lib/charty/backends/pyplot.rb
+++ b/lib/charty/backends/pyplot.rb
@@ -1,10 +1,10 @@
-require 'matplotlib/pyplot'
 require 'fileutils'
 
 module Charty
   module Backends
     class Pyplot
       def initialize
+        require 'matplotlib/pyplot'
         @plot = ::Matplotlib::Pyplot
       end
 
diff --git a/lib/charty/backends/rubyplot.rb b/lib/charty/backends/rubyplot.rb
index 26d0cae..dec5438 100644
--- a/lib/charty/backends/rubyplot.rb
+++ b/lib/charty/backends/rubyplot.rb
@@ -1,10 +1,10 @@
-require 'rubyplot'
 require 'fileutils'
 
 module Charty
   module Backends
     class Rubyplot
       def initialize
+        require 'rubyplot'
         @plot = ::Rubyplot
       end
 
```

I like your approach to handle errors and preparation of backends so I want to keep this direction.

Moreover, I have a test file I forgot to add the previous commit:

```ruby
require_relative "./test_helper"

class BackendsTest < Test::Unit::TestCase
  sub_test_case(".find_backend_class") do
    test("pyplot") do
      backend_class = Charty::Backends.find_backend_class(:pyplot)
      assert_equal("Charty::Backends::Pyplot", backend_class.name)

      backend_class = Charty::Backends.find_backend_class("pyplot")
      assert_equal("Charty::Backends::Pyplot", backend_class.name)
    end

    test("LoadError case") do
      assert do
        not Charty::Backends.names.include?("unable_to_load_backend")
      end
      assert_raise(Charty::BackendNotLoadedError) do
        Charty::Backends.find_backend_class(:unable_to_load_backend)
      end
      assert do
        not Charty::Backends.names.include?("unable_to_load_backend")
      end
    end
  end
end
```

I'll add some commits to add a test file like this, and fixations against your additional comments.

**Comment by mrkn** [comment]
I added a test case of Charty::Backends and performed a modification for allowing us to find a backend by string-like objects as Symbol is.

File: lib/charty/backends.rb [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+      @backends.keys
+    end
+
+    def self.register(name, backend_class)
+      case name
+      when Symbol
+        name = name.to_s
+      else
+        name = name.to_str
+      end
```
Comment (mrkn) [inline lib/charty/backends.rb]: @kou @284km 
Another option is only a Symbol is permitted as a backend name.
Which is better?

File: lib/charty/backends.rb [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+      @backends[name] = {
```
Comment (mrkn) [inline lib/charty/backends.rb]: Should we check the redefinition of the same backend name?

File: lib/charty/backends.rb (hunk 7 above)
Comment (kou) [reply lib/charty/backends.rb]: If we treat this as backend ID, `Symbol` only is better.
If we treat this as backend name, accepting `Symbol`, `String` and so on is better.

`:google_chart` looks ID not name. "Google Chart" looks name not ID.
(BTW, "Google Charts" (with the last "s") looks formal name.)
 
Anyway, this is not a strong opinion. I'm OK with the current behavior.
But could you extract normalization code as private `Backends.normalize_name` method or something?

File: lib/charty/backends.rb (hunk 8 above)
Comment (kou) [reply lib/charty/backends.rb]: We can do it but it's not important. Because backend developers don't do so.
If this is used by normal Charty users, it's better that this is implemented.

File: lib/charty/backends.rb (hunk 7 above)
Comment (mrkn) [reply lib/charty/backends.rb]: Done.

File: lib/charty/backends.rb (hunk 8 above)
Comment (mrkn) [reply lib/charty/backends.rb]: OK, I want to postpone the decision.

**Comment by mrkn** [comment]
I think all the things to do in this PR has been finished.
